File: LLD/Design-Patterns/ChainOfResponsibility/code/ChainOfResponsibility.cpp
// Chain of Responsibility — an expense request travels through TeamLead -> Manager -> Director.
// Compile: g++ -std=c++14 ChainOfResponsibility.cpp -o chain
// Run:     ./chain

#include <iostream>

class ApprovalHandler {
public:
    virtual void handle(double amount) = 0;
    virtual ~ApprovalHandler() = default;

    ApprovalHandler* setNext(ApprovalHandler* handler) {
        next_ = handler;
        return handler;
    }

protected:
    ApprovalHandler* next_ = nullptr;
};

class TeamLead : public ApprovalHandler {
public:
    void handle(double amount) override {
        if (amount <= 1000) {
            std::cout << "TeamLead approved expense of $" << amount << std::endl;
        } else if (next_) {
            std::cout << "TeamLead cannot approve $" << amount << " — escalating." << std::endl;
            next_->handle(amount);
        } else {
            std::cout << "No handler could approve $" << amount << std::endl;
        }
    }
};

class Manager : public ApprovalHandler {
public:
    void handle(double amount) override {
        if (amount <= 5000) {
            std::cout << "Manager approved expense of $" << amount << std::endl;
        } else if (next_) {
            std::cout << "Manager cannot approve $" << amount << " — escalating." << std::endl;
            next_->handle(amount);
        } else {
            std::cout << "No handler could approve $" << amount << std::endl;
        }
    }
};

class Director : public ApprovalHandler {
public:
    void handle(double amount) override {
        if (amount <= 20000) {
            std::cout << "Director approved expense of $" << amount << std::endl;
        } else if (next_) {
            next_->handle(amount);
        } else {
            std::cout << "No handler could approve $" << amount << " — needs CEO sign-off." << std::endl;
        }
    }
};

int main() {
    TeamLead teamLead;
    Manager manager;
    Director director;

    teamLead.setNext(&manager)->setNext(&director);

    std::cout << "Requesting approval for $500:" << std::endl;
    teamLead.handle(500);

    std::cout << "Requesting approval for $3000:" << std::endl;
    teamLead.handle(3000);

    std::cout << "Requesting approval for $18000:" << std::endl;
    teamLead.handle(18000);

    std::cout << "Requesting approval for $50000:" << std::endl;
    teamLead.handle(50000);

    return 0;
}

File: LLD/Design-Patterns/ChainOfResponsibility/code/chain_of_responsibility.py
"""
Chain of Responsibility — an expense request travels through TeamLead -> Manager -> Director.
Run: python chain_of_responsibility.py
"""

from abc import ABC, abstractmethod


class ApprovalHandler(ABC):
    def __init__(self):
        self.next = None

    def set_next(self, handler):
        self.next = handler
        return handler

    @abstractmethod
    def handle(self, amount):
        ...


class TeamLead(ApprovalHandler):
    def handle(self, amount):
        if amount <= 1000:
            print(f"TeamLead approved expense of ${amount}")
        elif self.next:
            print(f"TeamLead cannot approve ${amount} — escalating.")
            self.next.handle(amount)
        else:
            print(f"No handler could approve ${amount}")


class Manager(ApprovalHandler):
    def handle(self, amount):
        if amount <= 5000:
            print(f"Manager approved expense of ${amount}")
        elif self.next:
            print(f"Manager cannot approve ${amount} — escalating.")
            self.next.handle(amount)
        else:
            print(f"No handler could approve ${amount}")


class Director(ApprovalHandler):
    def handle(self, amount):
        if amount <= 20000:
            print(f"Director approved expense of ${amount}")
        elif self.next:
            self.next.handle(amount)
        else:
            print(f"No handler could approve ${amount} — needs CEO sign-off.")


if __name__ == "__main__":
    team_lead = TeamLead()
    manager = Manager()
    director = Director()

    team_lead.set_next(manager).set_next(director)

    print("Requesting approval for $500:")
    team_lead.handle(500)

    print("Requesting approval for $3000:")
    team_lead.handle(3000)

    print("Requesting approval for $18000:")
    team_lead.handle(18000)

    print("Requesting approval for $50000:")
    team_lead.handle(50000)
